### PR TITLE
chore: remove DbSlice mutex and add ConditionFlag in SliceSnapshot

### PR DESCRIPTION
**Pull request by kostasrim**
* remove DbSlice mutex
* add ConditionFlag in SliceSnapshot

Partially solves #4072

### CODE REVIEW

File: src/server/common.h [hunk 1]
```diff
@@ -355,4 +355,55 @@ template <typename Mutex> class ABSL_SCOPED_LOCKABLE SharedLock {
 
 extern size_t serialization_max_chunk_size;
 
+struct ConditionFlag {
+  util::fb2::CondVarAny cond_var;
+  bool flag = false;
+};
+
+// Helper class used to guarantee atomicity between serialization of buckets
+class ConditionGuard {
```
Comment (kostasrim) [inline src/server/common.h:364]: This is code we checked in and removed on previous PR's. I just brought it back

**Comment by kostasrim** [comment]
https://github.com/dragonflydb/dragonfly/actions/runs/11705220823

File: src/server/common.h (hunk 1 above)
Comment (dranikpg) [inline src/server/common.h:364]: How is this not just a thread local mutex? 🤔 Topological proof 🤓: you can move the constructor/destructor code into lock and unlock() of ConditionFlag and use it with lock_guard

File: src/server/common.h [hunk 2]
```diff
@@ -355,4 +355,55 @@ template <typename Mutex> class ABSL_SCOPED_LOCKABLE SharedLock {
...
+  ~ConditionGuard() {
+    enclosing_->flag = false;
+    enclosing_->cond_var.notify_one();
+  }
+
+ private:
+  ConditionFlag* enclosing_;
+};
+
+class LocalBlockingCounter {
```
Comment (dranikpg) [inline src/server/common.h:381]: we have EmbeddedBlockingCounter in helio, but it's atomicity is not needed here

File: src/server/snapshot.cc [hunk 3]
```diff
@@ -243,6 +243,7 @@ void SliceSnapshot::SwitchIncrementalFb(Context* cntx, LSN lsn) {
 }
 
 bool SliceSnapshot::BucketSaveCb(PrimeIterator it) {
+  ConditionGuard guard(&bucket_ser_);
```
Comment (adiholden) [inline src/server/snapshot.cc:246]: why here and in OnDbChange and not inside SerializeBucket ?

File: src/server/journal/streamer.h [hunk 4]
```diff
@@ -106,6 +106,8 @@ class RestoreStreamer : public JournalStreamer {
   cluster::SlotSet my_slots_;
   bool fiber_cancelled_ = false;
   bool snapshot_finished_ = false;
+
+  ConditionFlag bucket_ser_;
```
Comment (adiholden) [inline src/server/journal/streamer.h:110]: why did you add it to streamer?

Comment (kostasrim) [reply src/server/journal/streamer.h:110]: Because it has similar semantics with the `snapshot`. It traverses and serializes the buckets (see RestoreStreamer::Run)

File: src/server/snapshot.cc (hunk 3 above)
Comment (kostasrim) [reply src/server/snapshot.cc:246]: You have asked me the exact same question in June :rofl:

The answer is, that if you look on `BucketSaveCb` it calls `FlushChangeToEarlierCallbacks` which might preempt and when the fiber resumes the condition `the check` above might not be true. The code within `BucketSaveCb` should be atomic /considered as a critical section

File: src/server/common.h (hunk 2 above)
Comment (kostasrim) [reply src/server/common.h:381]: Yep that's why I introduced it

File: src/server/common.h (hunk 1 above)
Comment (kostasrim) [reply src/server/common.h:364]: :100:

File: src/server/db_slice.cc [hunk 5]
```diff
@@ -1158,6 +1143,11 @@ DbSlice::PrimeItAndExp DbSlice::ExpireIfNeeded(const Context& cntx, PrimeIterato
   const_cast<DbSlice*>(this)->PerformDeletion(Iterator(it, StringOrView::FromView(key)),
                                               ExpIterator(expire_it, StringOrView::FromView(key)),
                                               db.get());
+  // Replicate expiry
```
Comment (adiholden) [inline src/server/db_slice.cc:1146]: what is the reason you moved this here?

Comment (kostasrim) [reply src/server/db_slice.cc:1146]: Good question. The answer is `preemptions` because `heartbeat` is not `transactional`. If we first write to the journal, preempt and then delete, we risk another transaction coming in between the preemption and before the deletion. This is not good.

File: src/server/db_slice.cc [hunk 6]
```diff
@@ -815,10 +811,7 @@ void DbSlice::FlushDbIndexes(const std::vector<DbIndex>& indexes) {
 }
 
 void DbSlice::FlushDb(DbIndex db_ind) {
-  // We should not flush if serialization of a big value is in progress because this
-  // could lead to UB or assertion failures (while DashTable::Traverse is iterating over
-  // a logical bucket).
-  util::fb2::LockGuard lk(local_mu_);
+  block_counter_.Wait();
```
Comment (adiholden) [inline src/server/db_slice.cc:814]: is this true? we can not flush while we are in the middle of a big value serialization? 
when snapshoting we can run flushdb as we use intrusive ptr there

File: src/server/engine_shard.cc [hunk 7]
```diff
@@ -661,13 +661,17 @@ void EngineShard::Heartbeat() {
 
   CacheStats();
 
+  // TODO: iterate over all namespaces
+  DbSlice& db_slice = namespaces.GetDefaultNamespace().GetDbSlice(shard_id());
+  // Skip heartbeat if we are serializing a big value
+  if (db_slice.HasBlockingCounterMutating()) {
```
[the remark this answers is not in the document]
Comment (adiholden) [reply src/server/engine_shard.cc:667]: we should add some prints here so we will be able to know if hearbeat is not running for long time maybe because we are running lots of big values serialization or we have a bug somewere causing this skip

File: tests/dragonfly/instance.py [hunk 8]
```diff
@@ -124,6 +124,12 @@ def __init__(self, params: DflyParams, args):
             if threads > 1:
                 self.args["num_shards"] = threads - 1
 
+        if "disable_serialization_max_chunk_size" not in self.args:
+            # Add 1 byte limit for big values
+            self.args["serialization_max_chunk_size"] = 1
```
Comment (adiholden) [inline tests/dragonfly/instance.py:129]: does it realy makes sense to run always the big value serialization flow in regression tests?
it is good for extensive testing at the beging but then we dont test the other flow in code

File: src/server/db_slice.cc (hunk 6 above)
Comment (kostasrim) [reply src/server/db_slice.cc:814]: Well it did mess with `Traverse` in the past as it caused the assertion to trigger. Now we got bucket (not logical) Traversal it's probably ok. I would rather keep this though as a safe guard, and have another branch run this without this Wait here. We can incrementally remove those parts if you are ok with this

File: tests/dragonfly/replication_test.py [hunk 9]
```diff
@@ -2407,9 +2407,17 @@ async def test_replicate_old_master(
...
-    master = df_factory.create(version=1.19, path=released_dfly_path, cluster_mode=cluster_mode)
+    master = df_factory.create(
+        disable_serialization_max_chunk_size=0,
+        version=1.19,
+        path=released_dfly_path,
+        cluster_mode=cluster_mode,
+    )
     replica = df_factory.create(
-        cluster_mode=cluster_mode, cluster_announce_ip=announce_ip, announce_port=announce_port
+        disable_serialization_max_chunk_size=0,
```
[the remark this answers is not in the document]
Comment (kostasrim) [reply tests/dragonfly/replication_test.py:2417]: dfly_version = "v1.19.2" does not support the flag, so launching the instance will fail

File: tests/dragonfly/instance.py (hunk 8 above)
Comment (kostasrim) [reply tests/dragonfly/instance.py:129]: I wanted this to run for a week or so to make sure that we don't see any other new bugs. I will revert this afterwards. It was Roman's idea back in the summer :smile:

File: src/server/engine_shard.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (kostasrim) [reply src/server/engine_shard.cc:667]: +100 on this one. I thought the same tbh

File: tests/dragonfly/instance.py (hunk 8 above)
Comment (kostasrim) [reply tests/dragonfly/instance.py:129]: point is I either keep this PR and run the reg tests every 3 hours manually or I introduce this change and wait for a week :smile:

File: src/server/snapshot.cc (hunk 3 above)
Comment (adiholden) [reply src/server/snapshot.cc:246]: do you mean the check below lambda for version?
you can check also inside BucketSaveCb the snapshot version and if its bigger than dont serialize the bucket

File: src/server/journal/streamer.h (hunk 4 above)
Comment (adiholden) [reply src/server/journal/streamer.h:110]: we dont have any preeption in serializing buckets in RestoreStreamer as we do not support yet big value serialization in migration yet

File: tests/dragonfly/replication_test.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (adiholden) [reply tests/dragonfly/replication_test.py:2417]: in which version did you introduced this flag? maybe we can update this test to run this dfly version

File: src/server/db_slice.cc (hunk 6 above)
Comment (adiholden) [reply src/server/db_slice.cc:814]: I dont see how this is related to travesing on logical/ not logical bucket, we copy the pointer asside while flushing and snapshoting so that when flush is run we reset the pointer for the next transactions running.
Actually if we have a bug here I would like to understand before we merge this PR

File: src/server/db_slice.cc (hunk 5 above)
Comment (adiholden) [reply src/server/db_slice.cc:1146]: RecordExpiry can not preempt, the await is set to false
Does this assumption breaks now?

**Comment by adiholden** [comment]
It will be best if you add statistics for big value serilization i.e preept in serialize expose in stats
This way we will know this logic is executed in pytest flows and track this in running dragonfly servers

File: tests/dragonfly/replication_test.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (kostasrim) [reply tests/dragonfly/replication_test.py:2417]: > in which version did you introduced this flag?  

I will need to check.

> in which version did you introduced this flag? maybe we can update this test to run this dfly version

There is no point because a) the bug was on that version b) we won't run the reg tests with this flag on in the future anyway

[the remark this answers is not in the document]
Comment (adiholden) [reply tests/dragonfly/replication_test.py:2417]: I dont think should use this flag in regression tests so that small percentage of entries in replication/snapshot test will hit this flow.
The default for this flag will be much higher though

**Comment by kostasrim** [comment]
> It will be best if you add statistics for big value serilization i.e preept in serialize expose in stats This way we will know this logic is executed in pytest flows and track this in running dragonfly servers

Generally this PR is about  removing the lock from db slice and not adding statistics. I am happy to do this as part of this PR as it's a straightforward change but usually I like to separate tasks like that

File: tests/dragonfly/replication_test.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (kostasrim) [reply tests/dragonfly/replication_test.py:2417]: > I dont think should use this flag in regression tests so that small percentage of entries in replication/snapshot test will hit this flow.

Only for extreme testing. I can disable this no issue on that

File: src/server/db_slice.cc (hunk 6 above)
Comment (adiholden) [reply src/server/db_slice.cc:814]: they should not be freed by flushall, what is the different if you preempt from snapshoting in the middle of entry serialize or at the end of bucket serialize.
when running flushall the snapshot should continue to create a point in time snapshot for the data although flush all finised and next transactions get empty db
this is why we use intrusive ptr in snapshoting

Comment (kostasrim) [reply src/server/db_slice.cc:814]:
```
148 // We use reference counting semantics of DbTable when doing snapshotting.
149 // There we need to preserve the copy of the table in case someone flushes it during
150 // the snapshot process. We copy the pointers in StartSnapshotInShard function.
151 using DbTableArray = std::vector<boost::intrusive_ptr<DbTable>>;  
```

and 

```
   db_array_ = slice->databases();
```

Now it makes sense.

Comment (kostasrim) [reply src/server/db_slice.cc:814]: Something that pops my mind: 

Within FlushDbIndexes: 

```
   803                                                                                                   
   804   auto cb = [indexes, flush_db_arr = std::move(flush_db_arr)]() mutable {
   805     flush_db_arr.clear();
   806     ServerState::tlocal()->DecommitMemory(ServerState::kDataHeap | ServerState::kBackingHeap |
   807                                           ServerState::kGlibcmalloc);
   808   };
   809 
   810   fb2::Fiber("flush_dbs", std::move(cb)).Detach();

```

The fiber will clear `flush_db_arr` and then `call memory decommit`. This won't have any effect *if we are sharpshooing*. Shall we run `memory decommit` in the destructor instead ?

Comment (adiholden) [reply src/server/db_slice.cc:814]: yes we can

Comment (kostasrim) [reply src/server/db_slice.cc:814]: nice. It's minor but I will patch this at some point.

File: src/server/db_slice.h [hunk 10]
```diff
@@ -496,25 +495,18 @@ class DbSlice {
...
-      ABSL_LOCKS_EXCLUDED(local_mu_) {
-    util::fb2::LockGuard lk(local_mu_);
-    return pt->Traverse(cursor, std::forward<Cb>(cb));
-  }
-
   // Does not check for non supported events. Callers must parse the string and reject it
   // if it's not empty and not EX.
   void SetNotifyKeyspaceEvents(std::string_view notify_keyspace_events);
 
+  bool HasBlockingCounterMutating() const {
```
Comment (adiholden) [inline src/server/db_slice.h:502]: What we want to expose by this api is if we will preepmt if we try to do any mutation on db
i.e if heartbeat function want to do mutation for eviction/expiry and we do not allow preepmting there right now than is should invoke db slice api IsMutatingPreepmtive and if yes we will not invoke heartbeat eviction/expiry step
Note:  we should to call this in hearbeat and in PrimeEvictionPolicy::Evict 

I think that beside renaming the function name the logic will be more clear if we register 2 callbacks to DbSlice::RegisterOnChange one is the one we currently have the change cb, and the other one is the new callback that will return false or true if we will preempt if we invoke the callback. 
This way we will need to iterate though this callbacks to know if IsMutatingPreepmtive
On one hand it is not as simple as just returning the block_counter_ as you do now
But on the other hand I find this more robust and flow more clear when reading as it is more simple and clear from the register side what is this function and we will not need to remember to call the blocking_counter increamental from different places in code

File: src/server/common.h [hunk 11]
```diff
@@ -383,4 +383,32 @@ struct BorrowedInterpreter {
 
 extern size_t serialization_max_chunk_size;
```
Comment (adiholden) [inline src/server/common.h:384]: why serialization_max_chunk_size is global? can we just use the flag value in snapshot.cc?

File: src/server/server_family.cc [hunk 12]
```diff
@@ -2419,13 +2420,15 @@ void ServerFamily::Info(CmdArgList args, Transaction* tx, SinkReplyBuilder* buil
           current_snap_keys = res.current_keys;
           total_snap_keys = res.total_keys;
           perc = (static_cast<double>(current_snap_keys) / total_snap_keys) * 100;
+          big_value_preemptions = res.big_value_preemptions;
```
Comment (adiholden) [inline src/server/server_family.cc:2423]: I was thinking on a global counter similar to rdb_save_count
so that we can check this value even if snapshot is not in progress
use it in our py tests, check it is bigger than 0

File: src/server/common.h (hunk 11 above)
Comment (kostasrim) [reply src/server/common.h:384]: The reason is global is that we use it/cache the flag value in `ThreadLocalMutex`. 

1. When big value serialization is off, locking/unlocking the mutex is `noop`
2. When big value ser is on, locking/unlocking also checks that the same fiber does not lock the same mutex twice (it's for to detect deadlocks)

File: src/server/db_slice.h (hunk 10 above)
Comment (kostasrim) [reply src/server/db_slice.h:502]: As discussed internally, we can use this but for now I am aiming for correctness. Change should be simple though, we will discuss this offline.

File: src/server/rdb_save.cc [hunk 13]
```diff
@@ -159,7 +159,13 @@ std::string AbslUnparseFlag(dfly::CompressionMode flag) {
 }
 
 dfly::CompressionMode GetDefaultCompressionMode() {
-  return absl::GetFlag(FLAGS_compression_mode);
+  const auto flag = absl::GetFlag(FLAGS_compression_mode);
+  if (serialization_max_chunk_size == 0) {
+    return flag;
+  }
+
+  LOG_IF(WARNING, flag != dfly::CompressionMode::NONE);
```
Comment (adiholden) [inline src/server/rdb_save.cc:167]: you forgot to actually print something to log

File: src/server/dflycmd.cc [hunk 14]
```diff
@@ -120,6 +120,12 @@ void DflyCmd::ReplicaInfo::Cancel() {
     if (flow->cleanup) {
       flow->cleanup();
     }
+    // flow->cleanup() might be set when we transition to StableSync but the saver
```
Comment (adiholden) [inline src/server/dflycmd.cc:123]: I dont understand this comment
and its not related to big value serialization right?

Comment (kostasrim) [reply src/server/dflycmd.cc:123]: not really but we need to clean/reset the saver. Otherwise the destructor of `TheadLocalMutex` will complain

File: src/server/engine_shard.cc [hunk 15]
```diff
@@ -660,13 +660,24 @@ void EngineShard::Heartbeat() {
...
   CacheStats();
 
+  // TODO: iterate over all namespaces
+  DbSlice& db_slice = namespaces->GetDefaultNamespace().GetDbSlice(shard_id());
+  // Skip heartbeat if we are serializing a big value
+  static auto start = std::chrono::system_clock::now();
+  if (db_slice.WillBlockOnJournalWrite()) {
+    const auto elapsed = std::chrono::system_clock::now() - start;
+    if (elapsed > std::chrono::seconds(1)) {
+      LOG(WARNING) << "Stalled heartbeat() fiber for " << elapsed.count()
```
Comment (adiholden) [inline src/server/engine_shard.cc:670]: you should use log ever t
as if the heartbeat is stalled due to big value writings you we will be flooded with this prints

File: src/server/generic_family.cc [hunk 16]
```diff
@@ -597,6 +597,10 @@ void OpScan(const OpArgs& op_args, const ScanOpts& scan_opts, uint64_t* cursor,
   auto& db_slice = op_args.GetDbSlice();
   DCHECK(db_slice.IsDbValid(op_args.db_cntx.db_index));
 
+  // We need to make sure we don't preempt below, because we don't hold any locks to the keys
```
Comment (adiholden) [inline src/server/generic_family.cc:600]: I dont think its related to holding locks for keys
We do not allow preemption in traverse callback here, because other fiber can edit the bucket at the time we yield
Please write:
Because ScanCb can preempt due to journaling expired entries we need to make sure that we enter the callback in a timing when journaling will not cause preemtion

Comment (kostasrim) [reply src/server/generic_family.cc:600]: It's both. Let's ignore bucket splits when we preempt. Now let's say the bucket has 5 slots. 
We iterate the first and second and on the third we call `ExpireIfNeeded` from `ScanCb`. The issue is that we preempt `when we write to the journal` but we have `not` yet deleted the `entry` from the dash table. If we preempt on the first step, then we could get a transaction on the same key that just does a blind update. What's the problem there? That once we resume we will expire the `update`. That's a bug and its a side effect of not holding the locks to the relevant key.

Why for example it is ok to call `ExpireIfNeeded` and preempt in `Find` ? Well, because this is transactional and we hold the locks so even if we preempt we don't risk the corner case described above.

I will leave a big comment here.

File: src/server/db_slice.cc [hunk 17]
```diff
@@ -1098,6 +1090,7 @@ DbSlice::PrimeItAndExp DbSlice::ExpireIfNeeded(const Context& cntx, PrimeIterato
     LOG(ERROR) << "Invalid call to ExpireIfNeeded";
     return {it, ExpireIterator{}};
   }
+  block_counter_.Wait();
```
[the remark this answers is not in the document]
Comment (adiholden) [reply src/server/db_slice.cc:1093]: ?

File: src/server/generic_family.cc (hunk 16 above)
Comment (kostasrim) [reply src/server/generic_family.cc:600]: hmm I think I might be wrong because if we `preempt` when we write the journal change, the transaction that will try to modify the same key will block on `PreUpdate()` since it will try to call the registered callback that locks the same mutex and will block. In otherwords, the write to journal within ScanCb will `syncronize with PreUpdate()`

**Review by adiholden — APPROVED** [review]
